### ISSUE TITLE
How to copy bean 

### ISSUE THREAD

**Issue by all4you** [raphw/byte-buddy#1094]
Assuming that I have a Bean like `Source`:
```java
public class Source {
   private String name;
   private int age;
   public String getName() {
       return name;
   }
   public void setName(String name) {
       this.name = name;
   }
   public int getAge() {
       return age;
   }
   public void setAge(int age) {
       this.age = age;
   }
}
```
I want copy all the attributes from source bean to a target bean, like `Target`:

```java
public class Target {
   private String name;
   private int age;
   public String getName() {
       return name;
   }
   public void setName(String name) {
       this.name = name;
   }
   public int getAge() {
       return age;
   }
   public void setAge(int age) {
       this.age = age;
   }
}
```
How can I make the copy method with ByteBuddy:
```java
Target  target = copy(Source source, Target.class);
```

Thanks a lot!!

**Comment by raphw** [comment]
You can simply rename the class:

```java
new ByteBuddy().redefine(Source.class).name("Target").make();
```

Afterwards, you can choose to add additional properties or make further changes.

**Comment by all4you** [comment]
Sorry for the inappropriate demo, but I want to do the bean copy(or bean mapping) thing.

A more complicated deme maybe like this:
```java
public class Source {
   private String name;
   private int age;
   public String getName() {
       return name;
   }
   public void setName(String name) {
       this.name = name;
   }
   public int getAge() {
       return age;
   }
   public void setAge(int age) {
       this.age = age;
   }
   private Address address;
   public int getAddress() {
       return address;
   }
   public void setAddress(Address address) {
       this.address = address;
   }
}

public class Address {
   private String province;
   private String city;
   // getter、setter
}

public class Target {
   private String realName;
   private int realAge;
   public String getRealName() {
       return realName;
   }
   public void setRealName(String realName) {
       this.realName = realName;
   }
   public int getRealAge() {
       return realAge;
   }
   public void setRealAge(int realAge) {
       this.realAge = realAge;
   }
   private TargetAddress targetAddress;
   public int getTargetAddress() {
       return targetAddress;
   }
   public void setTargetAddress(TargetAddress targetAddress) {
       this.targetAddress = targetAddress;
   }
}

public class TargetAddress {
   private String province1;
   private String city1;
   // getter、setter
}
```

I want to copy all the attributes value from a source bean to a target bean like the lib `mapStruct` or `ModelMapper` do.
The `Target` class maybe has different attribute name from the `Source` Class.
I can customize the mapping if the attribute has different names:

Source.name ---> Target.realName
Source.age------> Target.realAge
Source.address--->Target.targetAddress
Source.address.province--->Target.targetAddress.province1
Source.address.city--->Target.targetAddress.city1

Or the mapping should be done automatically if the attribute has same type and name.

**Comment by raphw** [comment]
You would need to iterate over the fields of the classes you want to copy and [define the properties](https://github.com/raphw/byte-buddy/blob/master/byte-buddy-dep/src/main/java/net/bytebuddy/dynamic/DynamicType.java#L1163) for each field discovered. Byte Buddy then creates appropriate setters and getters.

If you discover a nested type you do want to emulate, you would create the same procedure recursively.

**Comment by all4you** [comment]
I know that  with `builder.defineProperty(name, type);` I can create the target class but the value is not assigned.
And what I need is not to create the class because the target class is exists, I need to set the value of the target instance. 

```java
Source source = new Source();
source.setName("foo");
source.setAge(20);
// I want copy the value from source to target so that I do not need do the set operation manually
Target target = copyValueFromSource(source, Target.class);
assertEquals(target.getName(), "foo");
```

**Comment by mariusoe** [comment]
Hi,

due to the fact that ByteBuddy focus on bytecode generation and transformation, I think that it is not the correct library to use to achieve your goal.

In case you don't want to do it manually, you could use `Gson`. Here, you can serialize your object into a JSON strucute, then deserialize it into your `Target` class. It's not the most performant way to do it, but it would work.

```
Gson gson= new Gson();
String tmp = gson.toJson(source);
Target targetObject = gson.fromJson(tmp, Target.class);
```

In order to solve the mapping problem, you can annotate the target's class properties (or the ones from the source class) which attribute should be used as value. See: https://howtodoinjava.com/gson/gson-serializedname/

**Comment by raphw** [comment]
That's indeed the argument for not offering such a specific implementation. The library is big as it is, it's up to users of Byte Buddy to create code for such mapping. You can for example use Jackson which even offers object to object mapping without an intermediate String serialization.

**Comment by all4you** [comment]
Thanks a lot for the suggestion, I have found several libs that do the mapping work, such as: 
- [dozer](https://github.com/DozerMapper/dozer) a mapping tool that copies data from one object to another using annotations and API or XML configuration
- [MapStruct](https://github.com/mapstruct/mapstruct) Code generator that simplifies mappings between different bean types, based on a convention-over-configuration approach
- [ModelMapper](https://github.com/modelmapper/modelmapper) Intelligent object mapping library that automatically maps objects to each other by many different convertors

I also found many examples and answers in stackoverflow but most of them are focused on code generation, to create a class, rather than create a bean or mapping a bean.

I just wondering is it possible to do the mapping work using Byte Buddy in an easy and grace way and also with high performance.  If it's possible then how can I do it, which api in ByteBuddy can I use to achieve the goal?
